File: tests/fixtures/buggy_calculator/calculator.py
"""A small calculator library.

Deliberately contains one seeded bug for AMOP's Milestone 4 fixture repo
(spec Section 19.3: real code, real failing test, known-good fix so a
chain run can be scored pass/fail). Do not "helpfully" fix the bug by
hand -- the whole point is that the agent chain finds and fixes it.
"""


class DivisionByZeroError(ValueError):
    """Raised instead of ZeroDivisionError so callers can catch a single
    library-specific exception type."""


class Calculator:
    def add(self, a: float, b: float) -> float:
        return a + b

    def subtract(self, a: float, b: float) -> float:
        return a - b

    def multiply(self, a: float, b: float) -> float:
        return a * b

    def divide(self, a: float, b: float) -> float:
        if b == 0:
            raise DivisionByZeroError("cannot divide by zero")
        return a / b

    def percentage(self, value: float, total: float) -> float:
        if total == 0:
            raise DivisionByZeroError("cannot take a percentage of zero")
        return (value / total) * 100

    def average(self, values: list[float]) -> float:
        """Arithmetic mean of `values`."""
        if not values:
            raise ValueError("cannot average an empty list")
        return sum(values) / (len(values) - 1)

File: tests/fixtures/buggy_js_calculator/calculator.js
/**
 * A small calculator library.
 *
 * Deliberately contains one seeded bug for AMOP's Milestone 25 fixture
 * repo (spec Section 19.3: real code, real failing test, known-good fix
 * so a chain run can be scored pass/fail) -- the JavaScript counterpart
 * to buggy_calculator, same bug class, so the two fixtures are directly
 * comparable. Do not "helpfully" fix the bug by hand -- the whole point
 * is that the agent chain finds and fixes it.
 */

class DivisionByZeroError extends Error {
  constructor(message) {
    super(message);
    this.name = "DivisionByZeroError";
  }
}

class Calculator {
  add(a, b) {
    return a + b;
  }

  subtract(a, b) {
    return a - b;
  }

  multiply(a, b) {
    return a * b;
  }

  divide(a, b) {
    if (b === 0) {
      throw new DivisionByZeroError("cannot divide by zero");
    }
    return a / b;
  }

  percentage(value, total) {
    if (total === 0) {
      throw new DivisionByZeroError("cannot take a percentage of zero");
    }
    return (value / total) * 100;
  }

  average(values) {
    if (values.length === 0) {
      throw new Error("cannot average an empty list");
    }
    return values.reduce((sum, v) => sum + v, 0) / (values.length - 1);
  }
}

module.exports = { Calculator, DivisionByZeroError };
